### ISSUE TITLE
Crash on using button "Sacrifice All" in Altar of Sacrifice

### ISSUE THREAD

**Issue by IvanSavenko** [vcmi/vcmi#2549]
**Describe the bug**
Seems to be reproducible crash when pressing "Sacrifice All" in Altar of Sacrifice.
You probably need some artifacts, I reproduced this bug with 2 equipped arts + 3 arts in the backpack.

**To Reproduce**
1. Start on map "Crimson and Clover" as good or neutral faction who can sacrifice arts
2. Use army & movement cheats
3. Collect some artifacts in vicinity, move some to backpack (might not be important)
4. Go to Altar of Sacrifice in the very center of the map
5. Click on "Sacrifice All" button

**Actual Behavior**
- Crash on Android
- Game hang on Linux
(probably iterator invalidation)

**Version**
 - OS: Android, Linux
 - Version: 1.3.0

**backtrace**
(CAltarWindow::putOnAltar(std::__ndk1::shared_ptr<CTradeWindow::CTradeableItem>, LIB_CLIENT::CArtifactInstance const*)+59)
(CAltarWindow::moveArtToAltar(std::__ndk1::shared_ptr<CTradeWindow::CTradeableItem>, LIB_CLIENT::CArtifactInstance const*)+1496)
(CAltarWindow::SacrificeAll()+1268)
(void LIB_CLIENT::CFunctionList<void ()>::operator()<>() const+1884)
(CButton::onButtonClicked()+168)
(EventDispatcher::handleLeftButtonClick(LIB_CLIENT::Point const&, bool)+196)
(InputSourceTouch::handleEventFingerUp(SDL_TouchFingerEvent const&)+184)
(InputHandler::processEvents()+87)
(CPlayerInterface::update()+1526)
(CGuiHandler::renderFrame()+123)

**Comment by SoundSSGood** [comment]
@IvanSavenko Very strange crash. Doesn't reproduce on Windows. But it is reproduced on Linux. Most similar to memory corruption CArtifactsOfHeroAltar::visibleArtSet::artifactsWorn.
The crash occurs when we try to erase a slot. (CArtHandler.cpp line 1075)
`artifactsWorn.erase(slot);`
Details of where and how the memory is corrupted so far I do not have.

**Comment by IvanSavenko** [comment]
@SoundSSGood 
Could it be iterator invalidation?
```
		for(const auto & aw : arts->visibleArtSet.artifactsWorn)
		{
			if(!aw.second.locked)
				moveArtToAltar(nullptr, aw.second.artifact);
		}
```
Code iterates over `artifactsWorn`, but (I assume that) `moveArtToAltar` eventually deletes current artifact from `artifactsWorn` invalidating iterator of this for loop.

Creating copy of `artifactsWorn` and iterating over it might help, but only if my guess is correct and not sure if this is correct solution here.

**Comment by IvanSavenko** [comment]
Fixed in #2556
